src/runtime/d3d12compute.cpp

```diff
@@ -0,0 +1,2794 @@
+
+    D3D12_CPU_DESCRIPTOR_HANDLE baseCPU = Call_ID3D12DescriptorHeap_GetCPUDescriptorHandleForHeapStart(descriptorHeap);
+    TRACEPRINT("descriptor heap base for CPU: " << baseCPU.ptr << "\n");
+    binder->CPU[UAV].ptr = (baseCPU.ptr += descriptorSize * 0);
+    binder->CPU[CBV].ptr = (baseCPU.ptr += descriptorSize * ResourceBindingLimits[UAV]);
+    binder->CPU[SRV].ptr = (baseCPU.ptr += descriptorSize * ResourceBindingLimits[CBV]);
+
+    D3D12_GPU_DESCRIPTOR_HANDLE baseGPU = Call_ID3D12DescriptorHeap_GetGPUDescriptorHandleForHeapStart(descriptorHeap);
+    TRACEPRINT("descriptor heap base for GPU: " << baseGPU.ptr << "\n");
+    binder->GPU[UAV].ptr = (baseGPU.ptr += descriptorSize * 0);
+    binder->GPU[CBV].ptr = (baseGPU.ptr += descriptorSize * ResourceBindingLimits[UAV]);
+    binder->GPU[SRV].ptr = (baseGPU.ptr += descriptorSize * ResourceBindingLimits[CBV]);
+
+    // initialize everything with null descriptors...
+    for (uint32_t i = 0; i < ResourceBindingLimits[UAV]; ++i) {
+        D3D12_UNORDERED_ACCESS_VIEW_DESC NullDescUAV = { };
+            NullDescUAV.Format = DXGI_FORMAT_R8G8B8A8_UNORM;  // don't care, but can't be unknown...
+            NullDescUAV.ViewDimension = D3D12_UAV_DIMENSION_BUFFER;
+            NullDescUAV.Buffer.FirstElement = 0;
+            NullDescUAV.Buffer.NumElements = 0;
+            NullDescUAV.Buffer.StructureByteStride = 0;
+            NullDescUAV.Buffer.CounterOffsetInBytes = 0;
+            NullDescUAV.Buffer.Flags = D3D12_BUFFER_UAV_FLAG_NONE;
+        D3D12_CPU_DESCRIPTOR_HANDLE hCPU = binder->CPU[UAV];
+        hCPU.ptr += i*descriptorSize;
+        (*device)->CreateUnorderedAccessView(NULL, NULL, &NullDescUAV, hCPU);
+    }
+    for (uint32_t i = 0; i < ResourceBindingLimits[CBV]; ++i) {
+        D3D12_CONSTANT_BUFFER_VIEW_DESC NullDescCBV = { };
+            NullDescCBV.BufferLocation = 0;
+            NullDescCBV.SizeInBytes = 0;
+        D3D12_CPU_DESCRIPTOR_HANDLE hCPU = binder->CPU[CBV];
+        hCPU.ptr += i*descriptorSize;
+        Call_ID3D12Device_CreateConstantBufferView((*device), &NullDescCBV, hCPU);
+    }
+    for (uint32_t i = 0; i < ResourceBindingLimits[SRV]; ++i) {
+        D3D12_SHADER_RESOURCE_VIEW_DESC NullDescSRV = { };
+            NullDescSRV.Format = DXGI_FORMAT_R8G8B8A8_UNORM;  // don't care, but can't be unknown...
+            NullDescSRV.ViewDimension = D3D12_SRV_DIMENSION_BUFFER;
+            NullDescSRV.Shader4ComponentMapping = D3D12_DEFAULT_SHADER_4_COMPONENT_MAPPING;
+            NullDescSRV.Buffer.FirstElement = 0;
+            NullDescSRV.Buffer.NumElements = 0;
+            NullDescSRV.Buffer.StructureByteStride = 0;
+            NullDescSRV.Buffer.Flags = D3D12_BUFFER_SRV_FLAG_NONE;
+        D3D12_CPU_DESCRIPTOR_HANDLE hCPU = binder->CPU[SRV];
+        hCPU.ptr += i*descriptorSize;
+        Call_ID3D12Device_CreateShaderResourceView((*device), NULL, &NullDescSRV, hCPU);
+    }
+
+    return binder;
+}
+
+struct NSRange {
```

Comment:
`NSRange` is an Apple Foundation type.  You don't need to use this; you can pass in the actual location/length.

Reply:
Ah, yes, that's a left-over from when I was trying to reuse the infrastructure of the metal back-end!